src/schedule/zephyr_ll.c

```diff
@@ -15,6 +15,12 @@
 
 #include <kernel.h>
 
+/* 1547fe68-de0c-11eb-8461-3158a1294853 */
+DECLARE_SOF_UUID("zll-schedule", zll_sched_uuid, 0x1547fe68, 0xde0c, 0x11eb,
+		 0x84, 0x61, 0x31, 0x58, 0xa1, 0x29, 0x48, 0x53);
+
+DECLARE_TR_CTX(ll_tr, SOF_UUID(zll_sched_uuid), LOG_LEVEL_INFO);
```

Comment:
Can you, please, rename `ll_tr` to something else, maybe `zll_tr`.
In i.MX case, we don't want to use `zephyr_ll` since this is limited to `TIMER_DOMAIN` (see [here](https://github.com/thesofproject/sof/blob/main/src/schedule/zephyr_ll.c#L450)) and we are using `DMA_DOMAIN`.
We want, for now, to keep using `ll_schedule`.

Therefore, when compiling this, for i.MX, I get: `zephyr_ll.c:22: multiple definition of ll_tr; `

Reply:
I'd like to help but you're asking me to fix a problem with a configuration that is currently not possible in this branch, probably only on your branch.  With this branch there is AFAIK no way to compile `ll_schedule.c` in Zephyr. After a fair amount of time spent realizing this I tried this naive:
```diff
--- a/zephyr/CMakeLists.txt
+++ b/zephyr/CMakeLists.txt
@@ -479,6 +479,7 @@ zephyr_library_sources(
 	${SOF_SRC_PATH}/schedule/schedule.c
 	${SOF_SRC_PATH}/schedule/dma_single_chan_domain.c
 	${SOF_SRC_PATH}/schedule/dma_multi_chan_domain.c
+	${SOF_SRC_PATH}/schedule/ll_schedule.c
 	${SOF_SRC_PATH}/schedule/zephyr_ll.c
 	${SOF_SRC_PATH}/schedule/zephyr.c
```
... but that does not work because:

```
sof/src/schedule/zephyr_ll.c:418: multiple definition of `zephyr_ll_task_init'; modules/sof/libmodules_sof.a(ll_schedule.c.obj):/p5/build/SOF/sof/src/schedule/ll_schedule.c:464: first defined here

sof/src/schedule/zephyr_ll.c:447: multiple definition of `zephyr_ll_scheduler_init'; modules/sof/libmodules_sof.a(ll_schedule.c.obj):/p5/build/SOF/sof/src/schedule/ll_schedule.c:624: first defined here
```

In other words @lyakh in PR #4377 made `zephyr_ll.c` and `ll_schedule.c` somehow mutually exclusive  even before this PR does with `ll_tr`.

Reply:
Forgot this: a simple rename and declaration of `DECLARE_TR_CTX(Zll_tr,...` in `zephyr_ll.c` is not possible because `ll_tr` is used in at least 5 different C files, some shared across both Zephyr and non-Zephyr.

Maybe I can figure out something else but I need to reproduce your problem first, so if you have a way to compile both `zephyr_ll.c` and `ll_schedule.c` please submit it first.

Reply:
@iuliana-prodan @dbaluta so DMA domain is becoming problematic as it does not scale well to multiple low latency streams (it's simple and great when there is a smaller number of streams though) or when the codec is driving the clocks.
I think we need to improve the DMA driven domain as follows.

Today we have

1) Zephyr LL gets its LL system tick from the **timer**. This runs in IRQ context and is non preemtable.
2) LL system tick wakes up high priority LL thread.
3) LL thread executes pipeline work.

We need to make one change to support DMA domain here and make it synchronous. Step 1 should also be able to be 

1) Zephyr LL gets its LL system tick from the **DMA IRQ**. This runs in IRQ context and is non preemtable.

This means we keep the DMA based time domain and we can also scale all LL streams to this time domain.

Fwiw, this is how things worked on OMAP4, the topology would tell FW which DMA channel was to be used as time domain. If the DMA channel stopped then FW had to then use the DMA channel from the next running stream. PR #4471 is allowing the trigger(start|stops) to be synchronised and this would mean that DMA IRQs could also be synchronised.

Reply:
@iuliana-prodan can you please:
- File a new issue about the longer term DMA domain discussion and solutions so @lgirdwood 's words of wisdom above do not get lost, buried deep down in this unrelated pull request
- Share in a draft PR showing how you compile `zephyr_ll.c` and `ll_schedule.c` together so I can reproduce the problem specific to your branch. It does not matter to me if that draft never gets merged, I just need a way to reproduce your problem if you would like me to solve it.

Reply:
@lgirdwood @marc-hb  I've created a pull request with my fixes for i.MX with Zephyr -see #4524 
Now we can run successfully playback and record scenarios with Zephyr.

@lgirdwood until we can extend `ll_zephyr` with DMA_IRQ (I've created a feature request for this - #4523) we'll be using `ll_schedule`. Hope it's ok with you.

@marc-hb using my pull request (#4524) and enabling dma-trace you'll run into `zephyr_ll.c:22: multiple definition of ll_tr;`
That's because in [commit 1](https://github.com/thesofproject/sof/pull/4524/commits/e1baf97460ce3809def6104e4ae073aae14bf243) I'm adding `ll_schedule` to be compiled for Zephyr and the `zephyr_ll` is added mandatory for all platforms - see [here](https://github.com/thesofproject/sof/blob/main/zephyr/CMakeLists.txt#L482).

Reply:
As explained in #4524 and #4525 it's not possible to compile-test any CONFIG_IMX yet but if I understood this issue correctly the latest version should be compatible with your branch, please test it.

Reply:
@iuliana-prodan I commented in your PR - maybe you shouldn't include both ll_zephyr.c and ll_schedule.c in your builds?

Reply:
Thanks, @lyakh.
I've replied to you in my PR.
Please let me know if you agree with my proposed solution - from [here](https://github.com/thesofproject/sof/pull/4524#discussion_r674588427), and I can make the changes and update that PR.